hugr-py/src/hugr/hugr.py

```diff
@@ -0,0 +1,313 @@
+
+        return o
+
+
+P = TypeVar("P", InPort, OutPort)
+
+
+def _unused_sub_offset(port: P, links: BiMap[OutPort, InPort]) -> P:
+    d: dict[OutPort, InPort] | dict[InPort, OutPort]
+    match port:
+        case OutPort(_):
+            d = links.fwd
+        case InPort(_):
+            d = links.bck
+    while port in d:
+        port = replace(port, sub_offset=port.sub_offset + 1)
+    return port
+
+
+@dataclass()
+class Hugr(Mapping[Node, NodeData]):
+    root: Node
+    _nodes: list[NodeData | None]
+    _links: BiMap[OutPort, InPort]
+    _free_nodes: list[Node]
+
+    def __init__(self, root_op: Op) -> None:
+        self._free_nodes = []
+        self._links = BiMap()
+        self._nodes = []
+        self.root = self.add_node(root_op)
+
+    def __getitem__(self, key: Node) -> NodeData:
+        try:
+            n = self._nodes[key.idx]
+        except IndexError:
+            n = None
+        if n is None:
+            raise KeyError(key)
+        return n
+
+    def __iter__(self):
+        return iter(self._nodes)
+
+    def __len__(self) -> int:
+        return self.num_nodes()
+
+    def add_node(
+        self,
+        op: Op,
+        parent: Node | None = None,
+    ) -> Node:
+        node_data = NodeData(op, parent)
+
+        if self._free_nodes:
+            node = self._free_nodes.pop()
+            self._nodes[node.idx] = node_data
+        else:
+            node = Node(len(self._nodes))
+            self._nodes.append(node_data)
+        return node
+
+    def delete_node(self, node: Node) -> NodeData | None:
+        for offset in range(self.num_in_ports(node)):
+            self._links.delete_right(node.inp(offset))
+        for offset in range(self.num_out_ports(node)):
+            self._links.delete_left(node.out(offset))
+
+        weight, self._nodes[node.idx] = self._nodes[node.idx], None
+        self._free_nodes.append(node)
+        return weight
+
+    def add_link(self, src: OutPort, dst: InPort) -> None:
+        src = _unused_sub_offset(src, self._links)
+        dst = _unused_sub_offset(dst, self._links)
+        if self._links.get_left(dst) is not None:
+            dst = replace(dst, sub_offset=dst.sub_offset + 1)
+        self._links.insert_left(src, dst)
+
+    def delete_link(self, src: OutPort, dst: InPort) -> None:
+        self._links.delete_left(src)
+
+    def num_nodes(self) -> int:
+        return len(self._nodes) - len(self._free_nodes)
+
+    def num_in_ports(self, node: Node) -> int:
+        return len(self.in_ports(node))
+
+    def num_out_ports(self, node: Node) -> int:
+        return len(self.out_ports(node))
+
+    def in_ports(self, node: Node) -> Collection[InPort]:
+        # can be optimised by caching number of ports
+        # or by guaranteeing that all ports are contiguous
+        return [p for p in self._links.bck if p.node == node]
+
+    def out_ports(self, node: Node) -> Collection[OutPort]:
+        return [p for p in self._links.fwd if p.node == node]
```

Comment:
Can we have a `Direction` enum, and and methods that choose input/output depending on the parameter?
Those ended up being quite useful on the rust implementation.